src/server/game/Movement/MovementGenerators/ChaseMovementGenerator.cpp

```diff
@@ -0,0 +1,182 @@
+    MovementGenerator const* gen = target->GetMotionMaster()->top();
+    if (gen->GetMovementGeneratorType() != CHASE_MOTION_TYPE)
+        return false;
+    return (static_cast<ChaseMovementGenerator const*>(gen)->GetTarget() == owner);
+}
+
+inline static float sq(float a) { return a*a; }
+static bool positionOkay(Unit* owner, Unit* target, Optional<float> minDistance, Optional<float> maxDistance, Optional<ChaseAngle> angle)
+{
+    float const distSq = owner->GetExactDistSq(target);
+    if (minDistance && distSq < sq(*minDistance))
+        return false;
+    if (maxDistance && distSq > sq(*maxDistance))
+        return false;
+    return !angle || angle->isAngleOkay(target->GetRelativeAngle(owner));
+}
+
+static const Optional<float> NONE;
+bool ChaseMovementGenerator::Update(Unit* owner, uint32 diff)
+{
+    // owner might be dead or gone (can we even get nullptr here?)
+    if (!owner || !owner->IsAlive())
+        return false;
+
+    // our target might have gone away
+    Unit* const target = GetTarget();
+    if (!target)
+        return false;
+
+    // the owner might've selected a different target (feels like we shouldn't check this here...)
+    if (owner->GetVictim() != target)
+        return false;
+
+    // the owner might be unable to move (rooted or casting), pause movement
+    if (owner->HasUnitState(UNIT_STATE_NOT_MOVE) || owner->IsMovementPreventedByCasting())
+    {
+
+        owner->StopMoving();
+        return true;
+    }
+
+    bool  const mutualChase     = isMutualChase(owner, target);
+    float const hitboxSum       = owner->GetCombatReach() + target->GetCombatReach();
+    float const minRange        = _range ? _range->minRange + hitboxSum : CONTACT_DISTANCE;
+    float const minTarget       = (_range ? _range->minTolerance : 0.0f) + hitboxSum;
+    float const maxRange        = _range ? _range->maxRange + hitboxSum : owner->GetMeleeRange(target); // melee range already includes hitboxes
+    float const maxTarget       = _range ? _range->maxTolerance + hitboxSum : CONTACT_DISTANCE + hitboxSum;
+    Optional<ChaseAngle> angle  = mutualChase ? Optional<ChaseAngle>() : _angle;
+
+    // if we're already moving, periodically check if we're already in the expected range...
+    if (owner->HasUnitState(UNIT_STATE_CHASE_MOVE))
+    {
+        if (_rangeCheckTimer > diff)
+            _rangeCheckTimer -= diff;
+        else
+        {
+            _rangeCheckTimer = RANGE_CHECK_INTERVAL;
+            if (positionOkay(owner, target, _movingTowards ? NONE : minTarget, _movingTowards ? maxTarget : NONE, angle))
+            {
+                owner->StopMoving();
+                _path = nullptr;
+            }
+        }
+    }
+
+    // if we're done moving, we want to clean up
+    if (owner->HasUnitState(UNIT_STATE_CHASE_MOVE) && owner->movespline->Finalized())
+    {
+        _path = nullptr;
+        owner->ClearUnitState(UNIT_STATE_CHASE_MOVE);
+
+        // @todo is this really the right place to call ::Attack?
```

Comment:
I've always had the same doubt, probably ``DoMeleeAttackIfReady`` is the best option, that is, if its even correct to call attack on reach

Reply:
`DoMeleeAttackIfReady` is on the AI, that makes even less sense.